File: Aplicação/backend/app/routers/portfolios.py
"""
Portfolio router for portfolio management endpoints.

Este router fornece endpoints para gestão completa de portfólios,
incluindo criação, atualização, exclusão e consulta de estatísticas.
"""

from typing import List, Optional
from fastapi import APIRouter, Depends, HTTPException, status, Query
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.deps import Session, CurrentUser
from app.schemas.portfolio import (
    Portfolio, PortfolioCreate, PortfolioUpdate, PortfolioSummary,
    PortfolioWithProjects
)
from app.services.portfolio_service import PortfolioService
# from app.core.exceptions import ValidationError, NotFoundError

router = APIRouter(
    prefix="/portfolios",
    tags=["portfolios"],
    responses={
        404: {"description": "Portfólio não encontrado"},
        400: {"description": "Dados inválidos"},
        401: {"description": "Não autorizado"},
        403: {"description": "Sem permissão"},
        500: {"description": "Erro interno do servidor"}
    }
)

@router.post(
    "/", 
    response_model=Portfolio, 
    status_code=status.HTTP_201_CREATED,
    summary="Criar novo portfólio",
    description="""
    Cria um novo portfólio para o usuário autenticado.
    
    **Validações:**
    - Nome do portfólio é obrigatório e único por usuário
    - Usuário deve estar autenticado
    - Descrição é opcional
    
    **Retorna:**
    - Portfólio criado com ID gerado automaticamente
    - Timestamps de criação e atualização
    """
)
async def create_portfolio(
    portfolio_data: PortfolioCreate,
    session: Session,
    current_user: CurrentUser
):
    """Criar um novo portfólio."""
    try:
        service = PortfolioService(session)
        portfolio = await service.create_portfolio(portfolio_data, current_user.id)
        return portfolio
    except ValueError as e:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail="Erro interno do servidor")

@router.get(
    "/", 
    response_model=List[PortfolioSummary],
    summary="Listar portfólios do usuário",
    description="""
    Lista todos os portfólios do usuário autenticado.
    
    **Parâmetros:**
    - include_inactive: Se deve incluir portfólios inativos (padrão: False)
    
    **Retorna:**
    - Lista de portfólios com informações resumidas
    - Contagem de projetos por portfólio
    - Status de ativação
    """
)
async def get_portfolios(
    session: Session,
    current_user: CurrentUser,
    include_inactive: bool = Query(False, description="Incluir portfólios inativos")
):
    """Listar todos os portfólios do usuário."""
    try:
        service = PortfolioService(session)
        portfolios = await service.get_user_portfolios(current_user.id, include_inactive)
        return portfolios
    except Exception as e:
        raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail="Erro interno do servidor")

@router.get(
    "/{portfolio_id}", 
    response_model=Portfolio,
    summary="Buscar portfólio por ID",
    description="""
    Busca um portfólio específico por ID.
    
    **Validações:**
    - Portfólio deve existir
    - Usuário deve ter permissão para acessar o portfólio
    
    **Retorna:**
    - Dados completos do portfólio
    - Timestamps de criação e atualização
    """
)
async def get_portfolio(
    portfolio_id: int,
    session: Session,
    current_user: CurrentUser
):
    """Buscar portfólio específico por ID."""
    try:
        service = PortfolioService(session)
        portfolio = await service.get_portfolio(portfolio_id, current_user.id)
        if not portfolio:
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Portfólio não encontrado")
        return portfolio
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail="Erro interno do servidor")

@router.put(
    "/{portfolio_id}", 
    response_model=Portfolio,
    summary="Atualizar portfólio",
    description="""
    Atualiza um portfólio existente.
    
    **Validações:**
    - Portfólio deve existir
    - Usuário deve ter permissão para atualizar o portfólio
    - Nome deve ser único por usuário (se alterado)
    
    **Retorna:**
    - Portfólio atualizado com novos dados
    - Timestamp de atualização modificado
    """
)
async def update_portfolio(
    portfolio_id: int,
    portfolio_data: PortfolioUpdate,
    session: Session,
    current_user: CurrentUser
):
    """Atualizar portfólio existente."""
    try:
        service = PortfolioService(session)
        portfolio = await service.update_portfolio(portfolio_id, portfolio_data, current_user.id)
        if not portfolio:
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Portfólio não encontrado")
        return portfolio
    except HTTPException:
        raise
    except ValueError as e:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail="Erro interno do servidor")

@router.delete(
    "/{portfolio_id}", 
    status_code=status.HTTP_204_NO_CONTENT,
    summary="Excluir portfólio",
    description="""
    Exclui um portfólio permanentemente.
    
    **Validações:**
    - Portfólio deve existir
    - Usuário deve ter permissão para excluir o portfólio
    
    **Atenção:**
    - Esta operação é irreversível
    - Todos os projetos do portfólio serão afetados
    """
)
async def delete_portfolio(
    portfolio_id: int,
    session: Session,
    current_user: CurrentUser
):
    """Excluir portfólio permanentemente."""
    try:
        service = PortfolioService(session)
        success = await service.delete_portfolio(portfolio_id, current_user.id)
        if not success:
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Portfólio não encontrado")
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail="Erro interno do servidor")

@router.get(
    "/{portfolio_id}/statistics",
    summary="Obter estatísticas do portfólio",
    description="""
    Obtém estatísticas detalhadas de um portfólio.
    
    **Retorna:**
    - Total de projetos
    - Projetos ativos e inativos
    - Status do portfólio
    - Timestamps de criação e atualização
    
    **Validações:**
    - Portfólio deve existir
    - Usuário deve ter permissão para acessar o portfólio
    """
)
async def get_portfolio_statistics(
    portfolio_id: int,
    session: Session,
    current_user: CurrentUser
):
    """Obter estatísticas detalhadas do portfólio."""
    try:
        service = PortfolioService(session)
        stats = await service.get_portfolio_statistics(portfolio_id, current_user.id)
        return stats
    except ValueError as e:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail="Erro interno do servidor")

@router.post(
    "/{portfolio_id}/activate",
    response_model=Portfolio,
    summary="Ativar portfólio",
    description="""
    Ativa um portfólio inativo.
    
    **Validações:**
    - Portfólio deve existir
    - Usuário deve ter permissão para ativar o portfólio
    
    **Retorna:**
    - Portfólio ativado com status atualizado
    """
)
async def activate_portfolio(
    portfolio_id: int,
    session: Session,
    current_user: CurrentUser
):
    """Ativar portfólio."""
    try:
        service = PortfolioService(session)
        portfolio = await service.activate_portfolio(portfolio_id, current_user.id)
        if not portfolio:
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Portfólio não encontrado")
        return portfolio
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail="Erro interno do servidor")

@router.post(
    "/{portfolio_id}/deactivate",
    response_model=Portfolio,
    summary="Desativar portfólio",
    description="""
    Desativa um portfólio ativo.
    
    **Validações:**
    - Portfólio deve existir
    - Usuário deve ter permissão para desativar o portfólio
    
    **Retorna:**
    - Portfólio desativado com status atualizado
    """
)
async def deactivate_portfolio(
    portfolio_id: int,
    session: Session,
    current_user: CurrentUser
):
    """Desativar portfólio."""
    try:
        service = PortfolioService(session)
        portfolio = await service.deactivate_portfolio(portfolio_id, current_user.id)
        if not portfolio:
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Portfólio não encontrado")
        return portfolio
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail="Erro interno do servidor")

File: Aplicação/frontend/src/api/portfolios.ts
import { useState } from 'react';
import api from './client';
import type { 
  Portfolio, 
  PortfolioCreate, 
  PortfolioUpdate, 
  PortfolioSummary,
  PortfolioWithProjects
} from '../types/portfolio';

// Serviços de API para portfólios
export const portfoliosApi = {
  // Listar todos os portfólios do usuário
  async getAll(includeInactive: boolean = false): Promise<PortfolioSummary[]> {
    const params = new URLSearchParams();
    if (includeInactive) {
      params.append('include_inactive', 'true');
    }
    
    const response = await api.get(`/portfolios?${params.toString()}`);
    return response.data;
  },

  // Buscar portfólio por ID
  async getById(id: number): Promise<Portfolio> {
    const response = await api.get(`/portfolios/${id}`);
    return response.data;
  },

  // Criar novo portfólio
  async create(data: PortfolioCreate): Promise<Portfolio> {
    const response = await api.post('/portfolios', data);
    return response.data;
  },

  // Atualizar portfólio
  async update(id: number, data: PortfolioUpdate): Promise<Portfolio> {
    const response = await api.put(`/portfolios/${id}`, data);
    return response.data;
  },

  // Deletar portfólio
  async delete(id: number): Promise<void> {
    await api.delete(`/portfolios/${id}`);
  },

  // Obter estatísticas do portfólio
  async getStats(id: number): Promise<any> {
    const response = await api.get(`/portfolios/${id}/statistics`);
    return response.data;
  },

  // Ativar portfólio
  async activate(id: number): Promise<Portfolio> {
    const response = await api.patch(`/portfolios/${id}/activate`);
    return response.data;
  },

  // Desativar portfólio
  async deactivate(id: number): Promise<Portfolio> {
    const response = await api.patch(`/portfolios/${id}/deactivate`);
    return response.data;
  },

  // Obter portfólio com projetos
  async getWithProjects(id: number): Promise<PortfolioWithProjects> {
    const response = await api.get(`/portfolios/${id}/projects`);
    return response.data;
  }
};

// Hook personalizado para portfólios
export const usePortfolios = () => {
  const [portfolios, setPortfolios] = useState<PortfolioSummary[]>([]);
  const [loading, setLoading] = useState(false);
  const [error, setError] = useState<string | null>(null);

  const fetchPortfolios = async (includeInactive: boolean = false) => {
    try {
      setLoading(true);
      setError(null);
      const data = await portfoliosApi.getAll(includeInactive);
      setPortfolios(data);
    } catch (err) {
      setError(err instanceof Error ? err.message : 'Erro ao carregar portfólios');
    } finally {
      setLoading(false);
    }
  };

  const createPortfolio = async (data: PortfolioCreate) => {
    try {
      setLoading(true);
      setError(null);
      const newPortfolio = await portfoliosApi.create(data);
      setPortfolios(prev => [...prev, newPortfolio]);
      return newPortfolio;
    } catch (err) {
      setError(err instanceof Error ? err.message : 'Erro ao criar portfólio');
      throw err;
    } finally {
      setLoading(false);
    }
  };

  const updatePortfolio = async (id: number, data: PortfolioUpdate) => {
    try {
      setLoading(true);
      setError(null);
      const updatedPortfolio = await portfoliosApi.update(id, data);
      setPortfolios(prev => prev.map(p => p.id === id ? updatedPortfolio : p));
      return updatedPortfolio;
    } catch (err) {
      setError(err instanceof Error ? err.message : 'Erro ao atualizar portfólio');
      throw err;
    } finally {
      setLoading(false);
    }
  };

  const deletePortfolio = async (id: number) => {
    try {
      setLoading(true);
      setError(null);
      await portfoliosApi.delete(id);
      setPortfolios(prev => prev.filter(p => p.id !== id));
    } catch (err) {
      setError(err instanceof Error ? err.message : 'Erro ao deletar portfólio');
      throw err;
    } finally {
      setLoading(false);
    }
  };

  const activatePortfolio = async (id: number) => {
    try {
      setLoading(true);
      setError(null);
      const activatedPortfolio = await portfoliosApi.activate(id);
      setPortfolios(prev => prev.map(p => p.id === id ? activatedPortfolio : p));
      return activatedPortfolio;
    } catch (err) {
      setError(err instanceof Error ? err.message : 'Erro ao ativar portfólio');
      throw err;
    } finally {
      setLoading(false);
    }
  };

  const deactivatePortfolio = async (id: number) => {
    try {
      setLoading(true);
      setError(null);
      const deactivatedPortfolio = await portfoliosApi.deactivate(id);
      setPortfolios(prev => prev.map(p => p.id === id ? deactivatedPortfolio : p));
      return deactivatedPortfolio;
    } catch (err) {
      setError(err instanceof Error ? err.message : 'Erro ao desativar portfólio');
      throw err;
    } finally {
      setLoading(false);
    }
  };

  return {
    portfolios,
    loading,
    error,
    fetchPortfolios,
    createPortfolio,
    updatePortfolio,
    deletePortfolio,
    activatePortfolio,
    deactivatePortfolio,
  };
};

// Hook para um portfólio específico
export const usePortfolio = (id: number) => {
  const [portfolio, setPortfolio] = useState<Portfolio | null>(null);
  const [loading, setLoading] = useState(false);
  const [error, setError] = useState<string | null>(null);

  const fetchPortfolio = async () => {
    try {
      setLoading(true);
      setError(null);
      const data = await portfoliosApi.getById(id);
      setPortfolio(data);
    } catch (err) {
      setError(err instanceof Error ? err.message : 'Erro ao carregar portfólio');
    } finally {
      setLoading(false);
    }
  };

  const updatePortfolio = async (data: PortfolioUpdate) => {
    try {
      setLoading(true);
      setError(null);
      const updatedPortfolio = await portfoliosApi.update(id, data);
      setPortfolio(updatedPortfolio);
      return updatedPortfolio;
    } catch (err) {
      setError(err instanceof Error ? err.message : 'Erro ao atualizar portfólio');
      throw err;
    } finally {
      setLoading(false);
    }
  };

  return {
    portfolio,
    loading,
    error,
    fetchPortfolio,
    updatePortfolio,
  };
};
